### PR TITLE
Separate out char marshalling rules to a separate MarshallingGeneratorFactory.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Separate marshalling for char out to a separate marshalling generator that allows us to always return the blittable marshaller for char collection elements.

This allows us to remove the AttributedMarshallingModel's special case for the char marshaller. Now it can exclusively special-case blittable marshalling and it doesn't need to know about char marshalling rules.

Also, fold in our custom "not supported" marshalling messages into the UnsupportedMarshallingFactory class so we can get the MarshalAsMarshallingGeneratorFactory closer to only supporting MarshalAs-style rules (no knowledge of the new attributed marshalling model)

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
Failure is https://github.com/dotnet/runtime/issues/70969

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/UnsupportedMarshallingFactory.cs [hunk 1]
```diff
@@ -12,7 +15,34 @@ public sealed class UnsupportedMarshallingFactory : IMarshallingGeneratorFactory
...
             throw new MarshallingNotSupportedException(info, context);
         }
+
+        public UnsupportedMarshallingFactory()
+            :this(DefaultTypeToErrorMessageMap)
+        {
+
+        }
+
+        public UnsupportedMarshallingFactory(ImmutableDictionary<ManagedTypeInfo, string> customTypeToErrorMessageMap)
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/UnsupportedMarshallingFactory.cs:34]:
```suggestion
        private UnsupportedMarshallingFactory(ImmutableDictionary<ManagedTypeInfo, string> customTypeToErrorMessageMap)
```

At least until we see a use case for it.

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/UnsupportedMarshallingFactory.cs:34]: I couldn't find one in this review.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/UnsupportedMarshallingFactory.cs [hunk 2]
```diff
@@ -12,7 +15,34 @@ public sealed class UnsupportedMarshallingFactory : IMarshallingGeneratorFactory
...
+        {
+            CustomTypeToErrorMessageMap = customTypeToErrorMessageMap;
+        }
+
+        public ImmutableDictionary<ManagedTypeInfo, string> CustomTypeToErrorMessageMap { get; }
+
+        public static ImmutableDictionary<ManagedTypeInfo, string> DefaultTypeToErrorMessageMap { get; } =
```
Comment (jkoritzinsky) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/UnsupportedMarshallingFactory.cs:41]:
```suggestion
        private static ImmutableDictionary<ManagedTypeInfo, string> DefaultTypeToErrorMessageMap { get; } =
```

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/CharMarshallingGeneratorFactory.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            }
+
+            // Explicit MarshalAs takes precedence over string encoding info
+            if (marshalInfo is MarshalAsInfo marshalAsInfo)
+            {
+                switch (marshalAsInfo.UnmanagedType)
+                {
+                    case UnmanagedType.I2:
+                    case UnmanagedType.U2:
+                        // If runtime marshalling is disabled, we treat UTF-16 char as blittable
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/CharMarshallingGeneratorFactory.cs:55]: Remove the comment? Not that you have separated runtime marshalling being disabled from this setting. Same below.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/UnsupportedMarshallingFactory.cs [hunk 4]
```diff
@@ -12,7 +15,34 @@ public sealed class UnsupportedMarshallingFactory : IMarshallingGeneratorFactory
...
+                throw new MarshallingNotSupportedException(info, context)
+                {
+                    NotSupportedDetails = errorMessage
+                };
+            }
             throw new MarshallingNotSupportedException(info, context);
         }
+
+        public UnsupportedMarshallingFactory()
+            :this(DefaultTypeToErrorMessageMap)
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/UnsupportedMarshallingFactory.cs:29]:
```suggestion
            : this(DefaultTypeToErrorMessageMap)
```

**Review by elinor-fung — APPROVED** [review]
Nice.

**Comment by jkoritzinsky** [comment]
Failures are #70969 and an intermittent NuGet failure.

**Pull request merged**